prinz/src/main/scala/pl/touk/nussknacker/prinz/util/converter/mlflow/MlflowDataConverter.scala

```diff
@@ -0,0 +1,61 @@
+package pl.touk.nussknacker.prinz.util.converter.mlflow
+
+import io.circe.{Decoder, Encoder, Json}
+import io.circe.generic.auto.{exportDecoder, exportEncoder}
+import io.circe.jawn.{parse, decode}
+import io.circe.syntax.EncoderOps
+import pl.touk.nussknacker.prinz.util.collection.immutable.SetMultimap
+
+class MlflowDataConverter {
+
+  implicit val encodeDecimalOrString: Encoder[Either[BigDecimal, String]] =
+    Encoder.instance(_.fold(_.asJson, _.asJson))
+
+  implicit val decodeDecimalOrString: Decoder[Either[BigDecimal, String]] =
+    Decoder[BigDecimal].map(Left(_)).or(Decoder[String].map(Right(_)))
+
+  case class Dataframe(columns: List[String], data: List[List[Either[BigDecimal, String]]]) {
+
+    def toList: List[(String, Either[BigDecimal, String])] =
+      (for (i <- columns.indices; record <- data) yield (columns(i), record(i))).toList
+  }
+
+  def toMultimap(data: String): SetMultimap[String, Either[BigDecimal, String]] = {
+    val dataframe = decode[Dataframe](parse(data).getOrElse(Json.Null).noSpaces).
```

Comment:
may be at least 2 lines of code to be more readable:
val parsedData = parse(data).getOrElse(Json.Null).noSpaces
val dataframe = ...

Reply:
watch out on dots at the end of the line